src/Microsoft.VisualStudio.Threading/AsyncSemaphore.cs

```diff
@@ -79,23 +100,40 @@ public Task<Releaser> EnterAsync(CancellationToken cancellationToken = default(C
                 return DisposedReleaserTask;
             }
 
-            return this.LockWaitingHelper(this.semaphore.WaitAsync(cancellationToken));
-        }
-
-        /// <summary>
-        /// Requests access to the lock.
-        /// </summary>
-        /// <param name="timeout">A timeout for waiting for the lock.</param>
-        /// <param name="cancellationToken">A token whose cancellation signals lost interest in the lock.</param>
-        /// <returns>A task whose result is a releaser that should be disposed to release the lock.</returns>
-        public Task<Releaser> EnterAsync(TimeSpan timeout, CancellationToken cancellationToken = default(CancellationToken))
-        {
-            if (this.disposed)
+            lock (this.syncObject)
```

Comment:
You are checking `disposed` before the lock, what if the variable is on the CPU cache and outdated? What if someone passed this check legitimately and is already waiting at the lock before the `Dispose` changes the variable? Should you be checking for disposal again after entering the lock?

This place is the only hole I see to make disposal more robust, it can lead to very hard to find bugs if a thread sneaks in between the dispose check and the lock. Since the fix is low complexity I think its worth it, even if you normally require callers to synchronize themselves for disposal (don't know what your assumptions are for shutdown).

Reply:
Ya, disposal is hard to make thread-safe in general. `SemaphoreSlim` itself isn't thread-safe for disposal. But I'd *love* for this class to be safe. I've made the change you recommended.